design/one-pager-class-selection.md

```diff
@@ -0,0 +1,371 @@
+used to select which nodes a pod may run on, which pods are endpoints of a
+service, and which Kubernetes cluster a Crossplane workload may run on.
+
+### Matching Scope
+
+Resource classes would be selected only from within the same namespace as the
+resource claim by default; this is the simplest, safest, and least surprising
+default for small scale Crossplane deployments. The infrastructure namespace
+pattern would be enabled by creating an association between application and
+infrastructure namespaces using namespace level annotations. For example:
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: acme-team
+  annotations:
+    resource-class.crossplane.io/select-namespaces: "acme-team, experimental-infra"
+```
+
+The above `acme-team` application namespace declares that any resource claims
+created within it should be matched against resource classes from namespaces
+`acme-team` or `experimental-infra`. Note that when the `select-namespaces`
+annotation exists it overrides the default behaviour of selecting only within
+the same namespace as the resource claim, and thus the namespace must declare
+itself to be in scope for class selection if it so desires. A special wildcard
+namespace `*` would allow matching across any namespace, for example:
+`resource-class.crossplane.io/select-namespaces: "*"`.
+
+### Exact Matching and Static Provisioning
+
+Resource claims would retain their existing `.spec.classRef` field under this
+proposal. In fact the `.spec.classSelector` would be used to set the `classRef`,
+just like Crossplane's [workload scheduling]. Resource claim authors who desired
+dynamic provisioning but had a specific resource class in mind could bypass the
+label selection process by explicitly referencing a resource class, for example:
+
+```yaml
+apiVersion: database.crossplane.io/v1alpha1
+kind: PostgreSQLInstance
+metadata:
+  namespace: acme-team
+  name: gitlab-database
+spec:
+  classRef:
+    apiVersion: database.gcp.crossplane.io/v1alpha2
+    kind: CloudSQLInstance
+    namespace: us-east-1
+    name: gitlab-experimental
+  writeConnectionSecretToRef:
+    name: app-postgresql-connection
+```
+
+Static provisioning is unaffected by this proposal; setting the
+`.spec.resourceRef` field of a resource claim would continue to bypass resource
+classes entirely. The resource claim would attempt to bind to the explicitly
+referenced managed resource.
+
+### Rudimentary Scheduling
+
+While not a goal of this proposal, it's worth noting that it begins to enable
+scheduling of resource claims to resource classes by decoupling the language
+used to match claims to classes from the (non-portable) language used to
+describe resource class parameters.
+
+It is challenging to match the requirements of a resource claim to the
+capabilities of a resource class on `spec` alone due to the "lowest common
+denominator" problem of multi cloud. The `RedisCluster` claim, for example,
```

Comment:
nit: could we add a link
https://thenewstack.io/avoiding-least-common-denominator-approach-hybrid-clouds/

Reply:
We should! I wanted to but couldn't immediately find an article that describes the concern as well as this one does. Thanks!